### PR TITLE
Refactor into "MemoryMap" class

### PR DESCRIPTION
**Pull request by IvanSanchez**
Refactor most functions into class methods - the "`Map` of `Uint8Array`s" was used so extensively that it makes sense to make it into class instances. Other functions are static methods of the class.

Changes the API surface completely, but (I think) for the better.

Also adds some rollup+buble bundling, `module` and `browser` in `package.json`, version bump to 1.0.0, the works.

### CODE REVIEW

File: intel-hex.js [hunk 1]
```diff
@@ -36,602 +33,618 @@ function checksumTwo(array1, array2) {
...
+     * @param {Iterable} blocks The initial value for the memory blocks inside this
+     * <tt>MemoryMap</tt>. All keys must be numeric, and all values must be instances of
+     * <tt>Uint8Array</tt>. Optionally it can also be a plain <tt>Object</tt> with
+     * only numeric keys.
+     */
+    constructor(blocks) {
+        this._blocks = new Map();
+
+        if (blocks && typeof blocks[Symbol.iterator] === 'function') {
+            this._blocks = new Map();
```
Comment (mrodem) [inline intel-hex.js]: Already done on line 87.

File: intel-hex.js [hunk 2]
```diff
@@ -36,602 +33,618 @@ function checksumTwo(array1, array2) {
...
+        hexLineRegexp.lastIndex = 0; // Reset the regexp, if not it would skip content when called twice
+
+        while ((matchResult = hexLineRegexp.exec(hexText)) !== null) {
+            recordCount++;
+
+            // By default, a regexp loop ignores gaps between matches, but
+            // we want to be aware of them.
+            if (lastCharacterParsed !== matchResult.index) {
+                throw new Error(
+                    'Malformed hex file: Could nor parse between characters ' +
```
Comment (mrodem) [inline intel-hex.js]: Typo.

File: intel-hex.js [hunk 3]
```diff
@@ -36,602 +33,618 @@ function checksumTwo(array1, array2) {
...
+                throw new Error('Checksum failed at record ' + recordCount + ' (' + matchResult[0].trim() + '), should be ' + cs.toString(16) );
+            }
 
-    return mergedBlocks;
-}
+            const offset = (recordBytes[1] << 8) + recordBytes[2];
+            const recordType = recordBytes[3];
+            let data = recordBytes.subarray(4);
 
+            if (recordType === 0) {
```
Comment (mrodem) [inline intel-hex.js]: Could the data record also be handled as part of the `switch`, similar to the other records?

File: intel-hex.js [hunk 4]
```diff
@@ -36,602 +33,618 @@ function checksumTwo(array1, array2) {
...
+                        break;
+
+                    case 5: // Start Linear Address Record
+                        // Do nothing. Record type 5 only applies to 32-bit Intel CPUs,
+                        // where it should reset the program counter (EIP CPU register)
+                        // It might have meaning for other CPU architectures
+                        // (see http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.faqs/ka9903.html )
+                        // but will be ignored nonetheless.
+                        break;
+                }
```
Comment (mrodem) [inline intel-hex.js]: Perhaps add a `default` here?

File: intel-hex.js [hunk 5]
```diff
@@ -36,602 +33,618 @@ function checksumTwo(array1, array2) {
...
+     */
+    paginate( pageSize=1024, pad=0xFF) {
+        if (pageSize <= 0) {
+            throw new Error('Page size must be greater than zero');
+        }
+    //     let pageAddr = -Infinity;
+        let outPages = new MemoryMap();
+        let page;
+
+        let sortedKeys = Array.from(this.keys()).sort((a,b)=>a-b);
```
Comment (mrodem) [inline intel-hex.js]: Consider extracting this to a function that returns sorted keys. All of `join`, `paginate`, and `asHexString` need this.

File: intel-hex.js [hunk 6]
```diff
@@ -36,602 +33,618 @@ function checksumTwo(array1, array2) {
...
+     *
+     * @param {Number} [pageSize=1024] The size of the output pages, in bytes
+     * @param {Number} [pad=0xFF] The byte value to use for padding
+     * @return {MemoryMap}
+     */
+    paginate( pageSize=1024, pad=0xFF) {
+        if (pageSize <= 0) {
+            throw new Error('Page size must be greater than zero');
+        }
+    //     let pageAddr = -Infinity;
```
Comment (mrodem) [inline intel-hex.js]: Remove commented out code.

File: intel-hex.js [hunk 7]
```diff
@@ -36,602 +33,617 @@ function checksumTwo(array1, array2) {
...
+    /**
+     * Returns a new instance of {@linkcode MemoryMap}, where:
+     *
+     * <ul>
+     *  <li>Each address is a multiple of <tt>pageSize</tt></li>
+     *  <li>The size of each <tt>Uint8Array</tt> is exactly <tt>pageSize</tt></li>
+     *  <li>Bytes from the input map to bytes in the output</li>
+     *  <li>Bytes not in the input are replaced by a padding value</li>
+     *  <li>If the optional start and end parameters are used, only bytes
+     *  between those addresses will be in the output.</li>
```
Comment (mrodem) [inline intel-hex.js]: JSDoc out of sync here perhaps? Cannot see any start and end parameters.

File: intel-hex.js [hunk 8]
```diff
@@ -36,602 +33,617 @@ function checksumTwo(array1, array2) {
...
+     *
+     * // Input can also be in an alternative syntax using a plain object instead of a Map
+     * let blocks = {
+     *      0: new Uint8Array(....),
+     *      0x01F0: new Uint8Array(....)
+     * };
+     *
+     * let string = arraysToHex(blocks);
+     */
+    asHexString(lineSize = 16) {
```
Comment (mrodem) [inline intel-hex.js]: JSDoc is out of sync here.

File: intel-hex.js [hunk 9]
```diff
@@ -648,57 +660,179 @@ function arraysToHex(blocks, lineSize = 16) {
...
+                    offsetRecord[4] = highAddress >> 24;    // new address offset, high byte
+                    offsetRecord[5] = highAddress >> 16;    // new address offset, low byte
 
                     records.push(
                         ':' +
-                        Array.prototype.map.call(recordHeader, hexpad).join('') +
-                        Array.prototype.map.call(subBlock, hexpad).join('') +
-                        hexpad(checksumTwo(recordHeader, subBlock))
+                        Array.prototype.map.call(offsetRecord, hexpad).join('') +
+                        hexpad(checksum(offsetRecord))
```
Comment (mrodem) [inline intel-hex.js:682]: I would have expected `offsetRecord.map(hexpad)` to work here, but apparently it does not?

File: intel-hex.js [hunk 10]
```diff
@@ -648,57 +660,179 @@ function arraysToHex(blocks, lineSize = 16) {
...
+                }
+
+                let recordSize = -1;
+                // Loop for every record for that spans the current 64KiB memory segment
+                while (lowAddress < 0x10000 && recordSize) {
+                    recordSize = Math.min(
+                        lineSize,                            // Normal case
+                        blockEnd - highAddress - lowAddress, // End of block
+                        0x10000 - lowAddress,                // End of low addresses
+                        255                                  // Maximum record length as per spec
```
Comment (mrodem) [inline intel-hex.js]: If max record length is given by the spec, then perhaps vaildate that `lineSize` is not larger than max?

**Review by mrodem — COMMENTED** [review]
Some general comments:
* Nice work! The API changes seem sensible.
* Readability of some functions could be improved, in particular `fromHex` and `asHexString`. I would perhaps consider creating separate `parseRecord` and `createRecord` functions for parsing/writing a single line of hex. It looks like you are prioritizing performance over readability, which is a valid trade off here. Try to avoid premature optimization though. :)
* I would recommend enabling ESLint while coding. We are using Airbnb's ESLint config in our other projects. I see a lot of cases where `const` could be used instead of `let`, unused variables, and other minor improvements.

**Comment by mrodem** [comment]
Oh, and by the way, :+1: for all the unit tests.

File: intel-hex.js (hunk 3 above)
Comment (IvanSanchez) [reply intel-hex.js]: Yes, *but* I decided to do it this way as to refactor the `offset===0` check. Doing it the current way saves writing that check five times.

File: intel-hex.js (hunk 7 above)
Comment (IvanSanchez) [reply intel-hex.js]: Apparently I planned that functionality but forgot to implement it in the end. The use case is having a different page size for the flash memory and for the UICR.

File: intel-hex.js (hunk 9 above)
Comment (IvanSanchez) [reply intel-hex.js:682]: Yeah, and that got me confused for a while too. Turns out that the `map` method of a `TypedArray` (a `Uint8Array` in this case) *returns a new `TypedArray`*. In this case, what we want is a `Uint8Array` as input, and an *`Array` of `String`s* as output. Hence the hack.

**Review by mrodem — APPROVED** [review]

**Pull request merged**